Fix split torch_function handling

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74727
* #83014
* #74728
* __->__ #83866

`Tensor.split` calls `TensorBase.split` whose `handle_torch_function` statement passes `func` as `Tensor.split` which is usually correct, but not here because of the use of `super()`. Instead this calls `torch._VF.split` which correctly differentiates from `torch.split`. This is currently okay since we never hit `TensorBase.split` for types with `__torch_function__` however, once we  allow skipping only one hop of `__torch_function__` this will expose the error.

peterbell10 committed:
Fix split torch_function handling

[ghstack-poisoned]

peterbell10 committed:
Update on "Fix split torch_function handling"




[ghstack-poisoned]

peterbell10 committed:
Update on "Fix split torch_function handling"




[ghstack-poisoned]

peterbell10 committed:
Update on "Fix split torch_function handling"




[ghstack-poisoned]

peterbell10 committed:
Fix RedispatchTensor serialization on "Fix split torch_function handling"




[ghstack-poisoned]

albanD commented on torch/_tensor.py:
```diff
@@ -763,16 +763,16 @@ def split(self, split_size, dim=0):
+                pass
+
+        if isinstance(split_size, int):
+            return torch._VF.split(self, split_size, dim)  # type: ignore[attr-defined]
```
Why not directly call `torch.split` here?

peterbell10 replied:
That wouldn't work because `torch.split` calls `Tensor.split`.

albanD replied:
Ho we expose split there from c++ and then we override it 5 lines later. Ok I missed that. Sounds good.

albanD reviewed (commented):
refactor sounds good.

peterbell10 committed:
Update on "Fix split torch_function handling"


`Tensor.split` calls `TensorBase.split` whose `handle_torch_function` statement passes `func` as `Tensor.split` which is usually correct, but not here because of the use of `super()`. Instead this calls `torch._VF.split` which correctly differentiates from `torch.split`. This is currently okay since we never hit `TensorBase.split` for types with `__torch_function__` however, once we  allow skipping only one hop of `__torch_function__` this will expose the error.


[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -l
